### PR TITLE
Update Scala to version 2.12.15

### PR DESCRIPTION
**Pull request by tgrospic**
## Overview

This PR is another update for Scala as preparation for Scala 2.13 and finally to Scala 3.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtName.scala [hunk 1]
```diff
@@ -25,21 +25,21 @@ object ListenAtName {
 
   private def buildPar[G[_]: BuildPar](name: G[Name]) = implicitly[BuildPar[G]].build(name)
 
-  implicit def buildParF[F[_]: MonadError[?[_], Throwable]] = new BuildPar[λ[A => F[Id[A]]]] {
+  implicit def buildParF[F[_]: MonadError[*[_], Throwable]] = new BuildPar[λ[A => F[Id[A]]]] {
```
Comment (zsluedem) [inline casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtName.scala:28]: What is the difference between `?` and `*`? 

I thought that it doesn't care which charactor it used.

File: project/Dependencies.scala [hunk 2]
```diff
@@ -106,30 +108,29 @@ object Dependencies {
...
+    "com.github.jnr"         % "jnr-ffi"     % "2.1.15",
+    "com.lihaoyi"            %% "geny"       % "0.6.10",
+    "com.lihaoyi"            %% "sourcecode" % "0.2.1",
+    "org.scala-lang.modules" %% "scala-xml"  % "1.3.0",
+    "com.typesafe"           % "config"      % "1.4.0"
   )
 
-  private val kindProjector = compilerPlugin("org.typelevel" %% "kind-projector" % "0.10.3")
+  private val kindProjector = compilerPlugin(
+    "org.typelevel" %% "kind-projector" % "0.13.2" cross CrossVersion.full
```
Comment (stanislavlyalin) [inline project/Dependencies.scala:132]: What does `cross CrossVersion.full` means? It is not clear for me from sbt documentation.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtName.scala (hunk 1 above)
Comment (stanislavlyalin) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtName.scala:28]: I guess, here is the [answer](https://github.com/scala/scala/pull/9560). But it is not clear for me. Look like `_` replaces with `?`, `?` replaces with `*` for a while, `*` replaces back to `_` again according to [the article](http://dotty.epfl.ch/docs/reference/changed-features/wildcards.html). What is the reason?

File: project/Dependencies.scala [hunk 3]
```diff
@@ -8,6 +8,7 @@ object Dependencies {
   val catsEffectVersion = "2.5.4"
   val catsMtlVersion    = "0.7.1"
   val fs2Version        = "2.5.10"
+  val monixVersion      = "3.4.0"
```
Comment (stanislavlyalin) [inline project/Dependencies.scala:11]: What is the reason to extract `monixVersion`? It is used only in one place.

**Review by zsluedem — APPROVED** [review]
LGTM. Left some questions.

Just curious. Maybe it would also make sense to upgrade to `2.13`? I think someone tries to do that before but failed.

File: models/src/main/scala/coop/rchain/models/ParMapTypeMapper.scala [hunk 4]
```diff
@@ -11,7 +11,10 @@ object ParMapTypeMapper {
 
   private[models] def parMapToEMap(parMap: ParMap): EMap =
     EMap(
-      parMap.ps.sortedList.map(t => zip(t._1, t._2)),
+      // Convert to Vector because with Scala 0.12.5 deserialized empty List (Nil)
```
Comment (stanislavlyalin) [inline models/src/main/scala/coop/rchain/models/ParMapTypeMapper.scala]: Did you mean `2.12.5`? It is similar to RNode version :slightly_smiling_face:

Comment (tgrospic) [reply models/src/main/scala/coop/rchain/models/ParMapTypeMapper.scala]: Thanks! Yes, I meant Scala version 2.12.15. :joy:

File: project/Dependencies.scala (hunk 3 above)
Comment (tgrospic) [reply project/Dependencies.scala:11]: Just to make it more explicit as it's important library for us and version is related to cats effect version.

File: project/Dependencies.scala (hunk 2 above)
Comment (tgrospic) [reply project/Dependencies.scala:132]: Sbt will use this information to correctly download (create download urls) artifacts from maven repository.

**Comment by tgrospic** [comment]
> LGTM. Left some questions.
> 
> Just curious. Maybe it would also make sense to upgrade to `2.13`? I think someone tries to do that before but failed.

@zsluedem Yes, multiple times. :) This https://github.com/tgrospic/rchain/pull/20 is my last attempt and I think I'm much closer now.

File: casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtName.scala (hunk 1 above)
Comment (tgrospic) [reply casper/src/main/scala/coop/rchain/casper/util/comm/ListenAtName.scala:28]: I think Scala folks just want to make thing more interesting. :smiley:

I think the main reason is to support partial application of type level functions.
```scala
F[_, B](A)  =>  F[A, B]
F[A, _](B)  =>  F[A, B]
```

**Review by stanislavlyalin — APPROVED** [review]
LGTM :+1:

**Comment by tgrospic** [comment]
bors merge

**Pull request merged**